lxd/instance/instance_utils.go

```diff
@@ -32,6 +32,9 @@ import (
 	"github.com/lxc/lxd/shared/version"
 )
 
+// ProfileValidationName is an indicator that the instance is just being used for profile validation.
+const ProfileValidationName = ""
```

Comment:
@stgraber I'm not really happy with this solution, although its not actually changing how it currently works in master branch, I've just formalised it with the use of a constant (with the view that I can easily find where it is used in the future to remove it!).

Instead I would like to do the following (but wanted to discuss it with you first before embarking on it):

- Create a new `instance.Instance` concrete implementation type called `Profile`
- This new concrete implementation type would satisfy the `instance.Instance` interface (albeit it most functions would return nil or not implemented error). However the functions related to config access would be implemented.
- It would also return an instance type of `instancetype.Any` indicating it is representing any instance type.
- This would then be used during profile validation when injected into the devices, allowing the devices to know they are doing profile validation and behave differently (allowing mixed VM and container configs).

- An optional step would be to move all of this new type implementation into a `common` type, and then have the `Profile` instance embed it.
- This would also allow us to embed the `common` type into `containerLXC` and `Qemu` types, allowing us to share some of the common functionality (like config expansion which is shared across all 3).

My only concern about this is that if I defined the new `common` type in the `instance` package, then it may be limited as to what it can import (and therefore limited as to what common functionality it can actually implement) because so many other packages do/will import the `instance` package, the likelihood for import loops is high.

I would prefer to have the `common` type in the same sub-package as `Qemu` and where `containerLXC` will eventually end up. However unlike our `device` and `storage` packages where all device types and driver types are defined in the same parent package , each instance type is being defined in its own package, e.g. `qemu` for VMs.

So I would propose moving the `Qemu` type into a package called perhaps `instances` or `drivers` (like storage) so that it is still separate from `instance` and `instancetype`, but so that eventually all 4 instance type implementations would end up in the same package (`Qemu`, `LXC`, `Profile` and `common`).

Ofcourse this will require containerLXC to be moved out of main, which requires the old storage layer to be removed entirely first.

Reply:
That seems reasonable.

Though reading this I was wondering, would it make sense to define a second interface used for config validation which would only include the subset we care about, reducing the need for empty functions in the profile case. Or do we need to pass that from function to function so much that it would be impractical?

Reply:
@stgraber so the things that `validateConfig()` inside the device's need are:

- inst.Name()
- inst.Type()
- inst.LocalDevices()
- inst.ExpandedDevices()

Validation also requires state to do DB lookups.

So I could create a new function to partially mimic `device.New` but only for validation called 

```go
device.New(inst instance.Instance, state *state.State, name string, conf deviceConfig.Device, volatileGet VolatileGetter, volatileSet VolatileSetter) (Device, error)
device.Validate(instConf instance.InstanceConfig, state *state.State, name string, conf deviceConfig.Device) error
```

This would accept the new interface `instance.InstanceConfig` rather than `instance.Instance`.

Then inside `Validate` it would instantiate a device as usual, but leave the instance, and volatile functions as nil.

Then I would modify the internal `d.validateConfig() error` function to accept the `instance.InstanceConfig` interface as an argument. It would then be able to use the subset of `instance.Instance` functions defined on that interface.

We would just need to make sure that nothing in `d.validateConfig()` ever access volatile keys or use `d.inst` internally (we would get a panic so would be obvious).

This would then retain the `device.New()` function never returning an invalid device.

I think there would still be value in moving `Qemu` type and the future `LXC` type into their own shared package and adding an internal `common` type so that we can move shared parts out of each implementation.

On a related topic, for both this and the devices generic vfs functions that are currently split over both the `drivers.common` type and the generic.go file, one way to be able to move them all into `drivers.common` type is to add a property to `drivers.common` that contains a point to the current driver, but as an interface, e.g.

go
```
type common struct {
	name        string
	config      map[string]string
	getVolID    func(volType VolumeType, volName string) (int64, error)
	commonRules *Validators
	state       *state.State
	logger      logger.Logger
	patches     map[string]func() error
        driver         driver // Add an interface pointer for the current driver inside the common struct
}
```

Then during driver load, we would pass the instantiated driver back into itself as `d.driver = d`.
E.g. https://play.golang.org/p/F1XCWJOVSyP

Reply:
Ah yeah, that's a good point, the main risk obviously is us accidentally using `d` when we mean `d.driver` and so using a generic function rather than the driver override one.

Reply:
@stgraber true. So for the generic vfs stuff there is an argument to say that these actually shouldnt be defined afainst the `common` struct anyway, because they aren't actually common to all drivers (as some drivers do not use the vfs functions). So moving all the vfs generic functions to a generic_vfs.go file (as some of them are now) makes sense.

So I'd be tempted to move them all out of `common`.

Then only functions that are truly common to all drivers will remain.

We could then do the same for the instance types, have a generic_instance.go file for functions are used in one or more of the instances, but need to access instance interface functions not implemented in `common`.

Reply:
Yeah, sounds reasonable